### PR TITLE
issue-20 to web-dev

### PR DESCRIPTION
**Pull request by sorianog**
Endpoints completed in this change:
POST /api/students/
GET /api/students/:userId
POST /api/students/:userId/drivingsessions
GET /api/students/:userId/drivingsessions

@happy25 Lmk if there is anything you want me to change. I also updated the Web API Endpoint doc in the Architecture Docs folder with some example requests/responses.  

*Note: I started working on other endpoints but I wanted to push up the ones that mobile needed first.  There are no implementations for processing driving data yet (i.e distance calculation). I am thinking of opening another issue for that one.*

### CODE REVIEW

File: web/app/routes/mainRoutes.js [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+var drivingSchoolCtrl = require("../controllers/drivingSchoolCtrl");
+var drivingSessionCtrl = require("../controllers/drivingSessionCtrl");
+
+// server routes ===========================================================
+// handle things like api calls
+// authentication routes
+
+// Get the JSON for the student with the specified _id
+router.get("/students/:_id", function(req, res) {
+  userCtrl.getStudent(req, res);
```
Comment (DylanWalseth) [inline web/app/routes/mainRoutes.js]: Not gonna lie, not a big fan of the whole request getting just passed a layer deeper for no real reason.
I think each of the gets/posts should handle direct req/res info and only pass the information as needed.
E.g.
```
router.get("/students/:_id", function(req, res) {
    userCtrl.getStudent(req.body, (student) => {
        res.statusCode = 201;
        res.json(student);
    }, (errMsg) => {
        res.statusCode = 400;
        res.send(errMsg);
    });
}
```

and
```
exports.createStudent = function(userInfo, callback, error) {
    var newStudent = new User({
        loginDetails: {
            userId: userInfo.loginDetails.userId,
            service: userInfo.loginDetails.service
        },
        firstName: userInfo.body.firstName,
        lastName: userInfo.body.lastName,
        userType: "student"
   });

    if(!userInfo.firstName){
        error("Missing first name");
    } else if(!userInfo.lastName) {
        error("Missing last name");
    } else {
        newStudent.save(function(err) {
            if(!err) {
                callback(newStudent);
            } else {
                error(`Error creating a new student, ${err}`);
            }
        });
    }
};
```
This will actually make the code much easier to test also.

File: web/app/routes/mainRoutes.js [hunk 2]
```diff
@@ -0,0 +1,83 @@
+var express = require("express");
+var router = express.Router();
+var mongoose = require("mongoose");
+mongoose.connect("mongodb://localhost/routerdb");
```
Comment (DylanWalseth) [inline web/app/routes/mainRoutes.js]: Leave this for now, but once you merge this in I'm going to move this to the server.js file and put the string into the config file I am building, considering we still don't have a production mongo server yet lol

Might also be worth adding that you need to have "Mongo server running locally required for development" to the README

File: web/app/routes/index.js [hunk 3]
```diff
@@ -0,0 +1,10 @@
+var express = require("express");
+var router = express.Router();
+
+// frontend routes =========================================================
+// route to handle all angular requests
+router.get("*", function(req, res) {
+  res.sendfile("./public/index.html");
+});
+
+module.exports = router;
```
Comment (DylanWalseth) [inline web/app/routes/index.js]: So, this confused for a while...

I saw this and was like "You aren't requiring this file from anywhere and adding it to the app.use. So this shouldn't work"
Then I tested it, and it worked.... which confused me ever so.

Turns out, this line saved you _partially_
```
// server.js line 27
app.use(express.static(__dirname + "/public"));
```
If you go to the homepage/root, it will default give you the index.html file, and since this is Angular, it will also give you every other view you need , such as the stats page, to make the Single Page App. Meaning you can press the stats page button and it will still work, but if you go straight to the /stats page, you would get a 404.
This confused for quite a bit

tl;dr fix
require and app.use this file in server.js

Comment (sorianog) [inline web/app/routes/index.js]: I have not touched the "server.js" during my work on the endpoints. Will this change have to be handled elsewhere?

Comment (sorianog) [inline web/app/routes/index.js]: Nvm, it looks like a small change so I can change it here.

File: web/app/routes/mainRoutes.js (hunk 1 above)
Comment (sorianog) [inline web/app/routes/mainRoutes.js]: So you are proposing that I separate out the content checking of the request and move this logic up a level outside of the "createStudent" method?

File: web/app/routes/mainRoutes.js (hunk 2 above)
Comment (sorianog) [inline web/app/routes/mainRoutes.js]: Added your statement to the README

File: web/app/controllers/drivingSessionCtrl.js [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    }
+  });
+}
+
+function checkCreateRequest(newDrivingSession, res) {
+  if(!newDrivingSession.startTime) {
+    res.status(400);
+    res.send("Start time required!");
+  } else if(!newDrivingSession.endTime){
+    res.status(400);
```
Comment (sorianog) [inline web/app/controllers/drivingSessionCtrl.js]: Yeah....I forgot to add the send....Now it is!

File: web/app/controllers/drivingSessionCtrl.js [hunk 5]
```diff
@@ -0,0 +1,84 @@
+var DrivingSession = require("../models/DrivingSession");
+var User = require("../models/User");
+
+//TODO: find an accurate formula for distance from gps points
```
Comment (sorianog) [inline web/app/controllers/drivingSessionCtrl.js]: I agree, we will have to look into this more

File: web/app/routes/mainRoutes.js (hunk 1 above)
Comment (DylanWalseth) [inline web/app/routes/mainRoutes.js]: Long story short, I was saying you should keep the req and res objects at this layer instead of passing them down to a lower layer

I mean in general, this
```
router.post("/students/", function(req, res) {
    userCtrl.createStudent(req, res);
});
```
seems kinda useless, I mean, why even have this function if all it does is just pass the entire req and res down to another function.
It's functional the same as just doing
```
router.post("/students/", userCtrl.createStudent);
```


To simplify to what I said in my previous comment (I also just noticed that I have the app.get function calling the wrong lower level function, opps. Same idea still stands though), I think you should do the whole "separation of concerns" thing. Don't have the function that has to get the user data also have to format the response. The example I gave above was that you request the certain user's info, then also give 2 functions, one if the user's info gathering was successful, another if there was an error. That way, the lower level function doesn't need to know how to format the response or what the functions even do (this makes testing significantly easier fyi) to handle it's own job, which should be to purely get the user's information.

File: web/app/controllers/drivingSchoolCtrl.js [hunk 6]
```diff
@@ -0,0 +1,42 @@
+var DrivingSchool = require("../models/DrivingSchool");
+
+exports.createSchool = function(req, res) {
```
Comment (DylanWalseth) [inline web/app/controllers/drivingSchoolCtrl.js]: You are repeating behavior that could be extracted out like so
```
exports.createSchool = function(req, res) {
    var requiredItems = ["addressLine1", "state", "zip"];
    var missingItems = GetMissingInformation(req.body, requiredItems);

    if (missingItems.length > 0) {
        res.status(400)
        res.json({
            "missing-items": missingItems
        });
        return;
    }
    var newSchool = new DrivingSchool({
        addressLine1: req.body.addressLine1,
        addressLine2: req.body.addressLine2,
        state: req.body.state,
        zip: req.body.zip
    });
    newSchool.save(function(err) {
        if (err) {
            res.status(400);
            res.json({"message":"Error creating a new school", "error": err});
        } else {
            res.status(201);
            res.json(newSchool);
        }
    });
}

function GetMissingInformation(jsonObject, keys) {
    var missingKeys = [];
    for (var key in keys) {
        if (!jsonObject.hasOwnProperty(keys[key]))
            missingKeys.push(keys[key]);
    }
    return missingKeys
}
```

Also, I would recommend always just responding with json, considering this is an api.

File: web/app/controllers/drivingSessionCtrl.js [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+function calculateDistance(drivePoints) {
+  // 0 for now
+  var distance = 0;
+  // drivePoints obj has lat, lon, speed, time
+  return distance;
+}
+
+function calculateDuration(startTime, endTime) {
```
Comment (DylanWalseth) [inline web/app/controllers/drivingSessionCtrl.js]: https://momentjs.com/docs/#/displaying/difference/

File: web/app/controllers/drivingSessionCtrl.js [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+  driveSessions = req.body;
+  driveSessions.forEach(function(driveSession) {
+    var finalDistance = calculateDistance(driveSession.DrivePoints);
+    var finalDuration = calculateDuration(driveSession.UnsyncDrive.startTime,
+                                            driveSession.UnsyncDrive.endTime);
+   newDrivingSession = new DrivingSession({
+      startTime: new Date(driveSession.UnsyncDrive.startTime),
+      endTime: new Date(driveSession.UnsyncDrive.endTime),
+      distance: finalDistance,
+      //TODO: figure out what units are
```
Comment (DylanWalseth) [inline web/app/controllers/drivingSessionCtrl.js]: Probably don't need to be more accurate than minutes, and at most, seconds.

File: web/app/routes/mainRoutes.js (hunk 1 above)
Comment (DylanWalseth) [inline web/app/routes/mainRoutes.js]: I also think this behavior should be reflected in several of the other files that do similar behavior to this.

**Comment by DylanWalseth** [comment]
Oh also, do `gulp lint` and cleanup all the errors before merging.

**Comment by DylanWalseth** [comment]
Another 2am thing to notice. Status code
You keep switching between `res.status(number)` and `res.statusCode = number` which behave the same, but should really be standardized for simplicity sake.

Also, very much worth noting that Express gives every response a default status code of 200, so for most of these, you are just rewriting over 200 with 200 FYI

**Comment by zawadzkip** [comment]
@DylanWalseth can we make a pre-commit script to run linting automatically for future web projects?

**Comment by DylanWalseth** [comment]
I made the changes I suggested and did a couple of other things to tidy up/simplify code, but also tried to keep the logic you had for the most part. Didn't touch the drivingSession file cause tbh, I am not sure what is happening there lol.

Also, @zawadzkip you can put the below into a file called `pre-commit` in the .git/hooks/ folder. I tried to get my SDL team last year to use it, and it really didn't happen as there is no way to force the client to run the command or to add this pre-commit file. But, the Travis build fails on linting error, so if someone pushes unlinted code, Travis will tell us.
```
#!/bin/sh
# Copy this into .git/hooks/pre-commit

# Lint stuff before committing
cd /web
gulp lint
```
This won't on Windows (I am pretty sure), but that shouldn't matter, as I think everyone doing web dev is using Linux.

**Comment by sorianog** [comment]
@DylanWalseth I was about to change stuff but you did it already. Got it. Mahalo!! <3 Do you need anything else from me?

File: web/app/controllers/drivingSchoolCtrl.js [hunk 9]
```diff
@@ -1,42 +1,47 @@
...
-      if(!err) {
-        res.statusCode = 201;
-        res.json(newSchool);
-      } else {
-        res.send("Error creating a new school, " + err);
-      }
-    });
-  }
+	if (missingItems.any()) {
+		error({
```
Comment (sorianog) [inline web/app/controllers/drivingSchoolCtrl.js]: Interesting syntax for error({}). I did not know that this can be done!

File: web/app/controllers/drivingSessionCtrl.js [hunk 10]
```diff
@@ -1,85 +1,91 @@
...
-      student.save(function(saveErr) {
-        if(saveErr){
-          res.send("Error saving driving session to student, " + saveErr);
-        }
-      });
-    } else {
-      res.send("Error retrieving driving session data, " + err);
-    }
-  });
+	User.findOne({
```
Comment (sorianog) [inline web/app/controllers/drivingSessionCtrl.js]: So this was changed structure/organization wise to make the code easier to read?

File: web/app/routes/mainRoutes.js [hunk 11]
```diff
@@ -1,24 +1,36 @@
 var express = require("express");
 var router = express.Router();
 var mongoose = require("mongoose");
+
+// For listing of supported status codes in this package
+// https://www.npmjs.com/package/http-status-codes
+var statusCodes = require("http-status-codes");
```
Comment (sorianog) [inline web/app/routes/mainRoutes.js]: Ah, I was thinking that there was a lib for status codes. I was just gonna straight up create my own static vars for em. It's great that there is an already existing one!

File: web/app/controllers/drivingSchoolCtrl.js [hunk 12]
```diff
@@ -3,9 +3,9 @@ var DrivingSchool = require("../models/DrivingSchool");
 // TODO: Check if Address, State, and Zip are valid (maybe some regex or external lib)
 exports.createSchool = function(info, callback, error) {
 	var requiredItems = ["addressLine1", "state", "zip"];
-	var missingItems = info.missingProperties(requiredItems);
+	var missingItems = MissingProperties(info, requiredItems);
 
-	if (missingItems.any()) {
+	if (_.any(missingItems)) {
```
Comment (sorianog) [inline web/app/controllers/drivingSchoolCtrl.js]: "_" is not underscore.js is it? I saw this in the code at work. lol

File: web/app/models/StateRegulations.js [hunk 13]
```diff
@@ -0,0 +1,17 @@
+// grab the mongoose module
```
Comment (sorianog) [inline web/app/models/StateRegulations.js]: I'm guessing this was merged in from Nate's branch for a closed issue?

File: web/app/routes/index.js [hunk 14]
```diff
@@ -3,8 +3,21 @@ var router = express.Router();
 
 // frontend routes =========================================================
 // route to handle all angular requests
-router.get("*", function(req, res) {
-	res.sendfile("./public/index.html");
+var staticFilesLocation = process.env.NODE_ENV === "production" ? "dist" : "public";
+router.get("/login", function(req, res) {
+	if(req.isAuthenticated()) {
```
Comment (sorianog) [inline web/app/routes/index.js]: Wow, thanks for adding the OAuth stuff in the endpoints!!! <3

File: web/karma.conf.js [hunk 15]
```diff
@@ -0,0 +1,69 @@
+// Karma configuration
```
Comment (sorianog) [inline web/karma.conf.js]: @DylanWalseth You got an recs for learning how to robustly test API? I asked someone at work for advice and will meet with this person in the future.

File: web/package.json [hunk 16]
```diff
@@ -5,14 +5,20 @@
...
+    "express-session": "^1.14.2",
     "method-override": "~2.0.2",
     "moment": "^2.17.0",
     "moment-timezone": "^0.5.10",
     "mongodb": "^2.2.10",
     "mongoose": "^4.1.12",
     "mongoose-auto-increment": "^5.0.1",
     "require-dir": "^0.3.1",
-    "underscore": "^1.8.3"
+    "underscore": "^1.8.3",
```
Comment (sorianog) [inline web/package.json]: Ah okay, disregard my previous comment. It is definitely underscore.js!

File: web/public/login.html [hunk 17]
```diff
@@ -0,0 +1,91 @@
+<!doctype html>
```
Comment (sorianog) [inline web/public/login.html]: Another merge from your branch @DylanWalseth?

File: web/server.js [hunk 18]
```diff
@@ -1,34 +1,210 @@
 // modules =================================================
-var express        = require("express");
-var app            = express();
-var bodyParser     = require("body-parser");
+var express		= require("express");
+var app			= express();
+var mongoose	   = require("mongoose");
+var bodyParser	 = require("body-parser");
 var methodOverride = require("method-override");
+var session = require("express-session");
```
Comment (sorianog) [inline web/server.js]: Mmmm, look at all those requires for OAuth. lol. It is definitely a beast on a whole different caliber!!

File: web/server.js [hunk 19]
```diff
@@ -1,34 +1,210 @@
...
 
+// If you don't have this file, contact Dylan for it.
+var config = require("./config.json");
+
 // config files
 var db = require("./config/db");
-var prodPort = 80;
-var devPort = 8080;
+
+var prodPort = config.ProdPort;
```
Comment (sorianog) [inline web/server.js]: Ah, so you can specify the production port that you want to use in the config file! Cool!

**Comment by sorianog** [comment]
Alright, so obvious the Travis build is failing because of the following as indicated in the build notes:

  39:14  error    No magic number: 400                         no-magic-numbers
  42:14  error    No magic number: 400                         no-magic-numbers
  47:16  error    No magic number: 201                         no-magic-numbers

Looks like it is dealing with the magic numbers for http status codes? Maybe there are some left behind and was not changed to use the lib that you added @DylanWalseth 

Anyways, after this is fixed, I think we can merge the changes, correct?

**Comment by DylanWalseth** [comment]
Still need to fix-up the driving session file. But I'm going to sleep. Too tired to try and figure that one out.

File: web/app/routes/index.js (hunk 14 above)
Comment (DylanWalseth) [inline web/app/routes/index.js]: This hasn't added it to the endpoints yet, only the frontend lol. We need to put the auth on the endpoints pretty much manually, tho, it should be easy

File: web/public/login.html (hunk 17 above)
Comment (DylanWalseth) [inline web/public/login.html]: Gonna be honest, not sure how this happened. git does some weird stuff with merging sometimes. I was just trying to catch this branch up to web-dev to get rid of merge conflicts

File: web/app/controllers/drivingSchoolCtrl.js [hunk 20]
```diff
@@ -0,0 +1,49 @@
+var DrivingSchool = require("../models/DrivingSchool");
+var _ = require("underscore");
+var of = require("../libs/objectFunctions");
+
+// TODO: Check if Address, State, and Zip are valid (maybe some regex or external lib)
+exports.createSchool = function(info, callback, error) {
+	var requiredItems = ["addressLine1", "state", "zip"];
+	var missingItems = of.MissingProperties(info, requiredItems);
```
Comment (sorianog) [inline web/app/controllers/drivingSchoolCtrl.js]: I can see this MissingProperties function being useful for other areas in our code. I'll see if I can use this as well!

File: web/app/controllers/drivingSessionCtrl.js [hunk 21]
```diff
@@ -0,0 +1,125 @@
+var DrivingSession = require("../models/DrivingSession");
+var User = require("../models/User");
+var _ = require("underscore");
+var geolib = require("geolib");
+var of = require("../libs/objectFunctions");
+var moment = require("moment"); // eslint-disable-line
+
+function calculateDistance(drivePoints) {
```
Comment (sorianog) [inline web/app/controllers/drivingSessionCtrl.js]: Wowza, @DylanWalseth You already implemented the distance calcs?! Sweeeet!

File: web/app/controllers/drivingSessionCtrl.js [hunk 22]
```diff
@@ -0,0 +1,125 @@
...
+			fun(a);
+		}
+	};
+}
+
+// Welcome to callback hell
+// mwhahahahaah
+// seriously tho, I will comeback and clean this up, got all these async calls tho
+// Ever tried to do an array of async calls? Not fun.
+// pinky promise @dylanwalseth
```
Comment (sorianog) [inline web/app/controllers/drivingSessionCtrl.js]: Lol, I have witnessed and read your "pink promise". There's no going  back on this now!!! Muwahaha

File: web/app/models/DrivingSchool.js [hunk 23]
```diff
@@ -1,18 +1,47 @@
...
 var DrivingSchoolSchema = new Schema({
-    schoolId: { type: Number, required: true },
-    owners: [{ type: Schema.Types.ObjectId, ref: "User" }],
-    instructors: [{ type: Schema.Types.ObjectId, ref: "User" }],
-    students: [{ type: Schema.Types.ObjectId, ref: "User" }],
-    addressLine1: { type: String, required: true },
-    addressLine2: { type: String, required: true },
-    state: { type: String, required: true },
-    zip: { type: Number, required: true }
+	schoolId: {
```
Comment (sorianog) [inline web/app/models/DrivingSchool.js]: Yep, definitely more readable here. Thanks for changing this!!

File: web/spec/app/controllers/drivingSchoolCtrlSpec.js [hunk 24]
```diff
@@ -0,0 +1,91 @@
+var DrivingSchool = require(`${process.env.PWD}/app/models/DrivingSchool`);
```
Comment (sorianog) [inline web/spec/app/controllers/drivingSchoolCtrlSpec.js]: Ah, so that is what you were talking about that last time regarding the syntax needed to determine the current working directory!
